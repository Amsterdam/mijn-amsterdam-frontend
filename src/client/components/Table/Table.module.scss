@import '../../styles/_global.scss';
@import '../Button/Button.module.scss';

.Table {
  width: 100%;
  border: 0;
  border-collapse: collapse;

  tr:first-of-type td {
    border-top: 0;;
  }

  @include mq-phone() {
    display: grid;
    tr,
    td,
    tbody {
      display: block;
    }
    td:not(:first-child) {
      display: block;
    }
    thead {
      display: none;
    }
    tr:last-of-type td {
      border-bottom: 0;
    }
  }
  thead th {
    border-bottom: 0.1rem solid $color-neutral-grey2;
    padding-bottom: .5rem;
  }
}

.TableRow {
  td {
    padding: 1em 1em 1em 0;
    vertical-align: top;
    border-top: 0.1rem solid $color-neutral-grey2;
  }
  th {
    text-align: left;
    color: $color-neutral-grey4;
    font-weight: 500;
  }
  @include mq-phone() {
    padding-top: .5rem;
    border-top: 0.1rem solid $color-neutral-grey2;
    td {
      border-top: 0;
      padding: 0;
      &:last-child {
        padding-bottom: 1rem;
      }
    }
  }
}

.DisplayPropTitle {
  @include select(ButtonLink_ButtonLink) {
    white-space: normal;
    display: block;
  }
}
.DisplayPropLabel {
  display: none;
  padding-right: 0.3rem;

  @include mq-phone() {
    display: initial;
  }
}
.DisplayProp {
  color: $color-neutral-grey4;
  font-weight: normal;
  white-space: nowrap;
}
