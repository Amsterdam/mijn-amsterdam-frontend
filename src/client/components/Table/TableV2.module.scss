.TableV2 {
  width: 100%;
  margin-bottom: var(--ams-grid-row-gap-sm);

  :global {
    .ams-table__caption {
      font-size: var(--ams-text-level-2-font-size);
    }

    tr {
      .ams-table__cell,
      .ams-table__header-cell {
        hyphens: var(--ams-hyphenation-hyphens);
        &:first-child {
          padding-inline-start: 0 !important;
        }
        &:last-child {
          padding-inline-end: 0 !important;
        }
      }
    }
  }
}
