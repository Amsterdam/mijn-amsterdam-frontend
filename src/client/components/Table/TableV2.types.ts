import { ReactNode } from 'react';

import { Unshaped, type LinkProps } from '../../../universal/types/App.types';

export type DisplayPropsBase<
  T,
  T2 = { detailLinkComponent: string } & T,
> = Readonly<{
  [Property in keyof T2]+?: string | number | ReactNode;
}>;

export type DisplayProps<T> =
  | DisplayPropsBase<T>
  | {
      props: DisplayPropsBase<T>;
      // colwidths can be used to set the width of each column in the table.
      // the array always has the size of the number of keys in the props object. e.g. if props has 4 keys, colWidths will have 4 values.
      // if a colWidth value is '0', the column will not be displayed and prop at entries(props) will be discarded. e.g.
      // props={ foo: 'bar', baz: 'world' } colWidths={{ large: ['50%', '50%'], small: ['100%', '0'] }} will only display the foo column on small screens and both columns on large screens.
      colWidths?: TableV2ColWidths;
    };

export type ScreenSize = 'small' | 'large';
export type TableV2ColWidths = Record<ScreenSize, string[]>;

export interface TableV2Props<T> {
  displayProps: DisplayProps<T>;
  items: T[];
  className?: string;
  showTHead?: boolean;
  caption?: string;
  subTitle?: ReactNode;
}

export interface ObjectWithOptionalLinkAttr extends Unshaped {
  link?: LinkProps;
}

export type WithDetailLinkComponent<T> = T & {
  detailLinkComponent?: ReactNode;
};
