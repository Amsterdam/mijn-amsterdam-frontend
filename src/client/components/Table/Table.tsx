import classnames from 'classnames';
import React, { ReactNode } from 'react';
import { capitalizeFirstLetter, entries } from '../../../universal/helpers';
import { Unshaped } from '../../../universal/types';
import Linkd from '../Button/Button';
import styles from './Table.module.scss';
import InnerHtml from '../InnerHtml/InnerHtml';

export function addTitleLinkComponent(
  items: Unshaped[],
  titleKey: string = 'title'
) {
  return items.map((item: any) => {
    return {
      ...item,
      [titleKey]: (
        <Linkd href={item.link.to}>
          {capitalizeFirstLetter(item[titleKey]) || 'uhhhh'}
        </Linkd>
      ),
    };
  });
}

export interface TableProps {
  items: Unshaped[];
  className?: string;
  titleKey?: string;
  displayProps?: { [key: string]: string | number | ReactNode };
}

export default function Table({
  items,
  displayProps,
  titleKey = 'title',
  className,
}: TableProps) {
  const displayPropsFinal = !displayProps ? { [titleKey]: ' ' } : displayProps;
  const displayPropEntries = entries(displayPropsFinal).filter(
    ([key]) => key !== titleKey
  );
  const hasDisplayPropTableHeadingLabels = !!Object.keys(displayPropsFinal)
    .length;

  return (
    <table className={classnames(styles.Table, className)}>
      {hasDisplayPropTableHeadingLabels && (
        <thead>
          <tr className={styles.TableRow}>
            {!!items[0] && titleKey in items[0] && (
              <th className={styles.DisplayProp}>
                {displayPropsFinal[titleKey] || ' '}
              </th>
            )}
            {displayPropEntries.map(([key, label]) => (
              <th key={`th-${key}`} className={styles.DisplayProp}>
                {label}
              </th>
            ))}
          </tr>
        </thead>
      )}
      <tbody>
        {items.map((item: Unshaped, index) => (
          <tr
            key={item.id || `${titleKey}-${index}`}
            className={styles.TableRow}
          >
            {titleKey in item && (
              <td className={styles.DisplayPropTitle}>{item[titleKey]}</td>
            )}
            {displayPropEntries.map(([key, label]) => (
              <td key={`td-${key}`} className={styles.DisplayProp}>
                {!!label && (
                  <span className={styles.DisplayPropLabel}>{label}:</span>
                )}
                {item[key] ? (
                  <InnerHtml>{item[key]}</InnerHtml>
                ) : (
                  <span>&mdash;</span>
                )}
              </td>
            ))}
          </tr>
        ))}
      </tbody>
    </table>
  );
}
