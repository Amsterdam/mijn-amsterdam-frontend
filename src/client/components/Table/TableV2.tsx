import { Link } from '@amsterdam/design-system-react';
import classnames from 'classnames';
import { ReactNode } from 'react';
import { capitalizeFirstLetter, entries } from '../../../universal/helpers';
import { LinkProps, Unshaped } from '../../../universal/types';
import styles from './TableV2.module.scss';

interface ObjectWithOptionalLinkAttr extends Unshaped {
  link?: LinkProps;
}

interface ObjectWithOptionalId extends Unshaped {
  id?: number | string;
}

export function addTitleLinkComponent<T extends ObjectWithOptionalLinkAttr>(
  items: T[],
  titleKey: keyof T = 'title'
) {
  return items.map((item) => {
    if (!item.link?.to) {
      return item;
    }

    let title: string = item[titleKey];
    if (typeof title !== 'string') {
      title = 'Onbekend item';
    }

    return {
      ...item,
      [titleKey]: (
        <Link href={item.link.to}>{capitalizeFirstLetter(title)}</Link>
      ),
    };
  });
}

export type DisplayProps<T> = {
  [Property in keyof T]+?: string | number | ReactNode;
};

export interface TableV2Props<T> {
  items: T[];
  className?: string;
  displayProps: DisplayProps<T> | null;
  gridColStyles?: string[];
  showTHead?: boolean;
}

export function TableV2<T extends ObjectWithOptionalId>({
  items,
  displayProps,
  className,
  gridColStyles,
  showTHead = true,
}: TableV2Props<T>) {
  const displayPropEntries = displayProps !== null ? entries(displayProps) : [];

  return (
    <table
      className={classnames(
        styles.Table,
        !!gridColStyles?.length && styles.Table__grid,
        className
      )}
    >
      {showTHead && (
        <thead>
          <tr
            className={classnames(
              styles.TableRow,
              !!gridColStyles?.length && [
                'amsterdam-grid',
                styles.TableRow__Grid,
              ]
            )}
          >
            {displayPropEntries.map(([key, label], index) => {
              const HeadingCellElement = !!label ? 'th' : 'td';
              return (
                <HeadingCellElement
                  key={`th-${key}`}
                  className={gridColStyles?.[index]}
                >
                  {label}
                </HeadingCellElement>
              );
            })}
          </tr>
        </thead>
      )}
      <tbody>
        {items.map((item, index) => (
          <tr
            key={item.id ?? `tr-${index}`}
            className={classnames(
              styles.TableRow,
              !!gridColStyles?.length && [
                'amsterdam-grid',
                styles.TableRow__Grid,
              ]
            )}
          >
            {displayPropEntries.map(([key, label], index) => (
              <td key={`td-${key}`} className={gridColStyles?.[index]}>
                {item[key]}
              </td>
            ))}
          </tr>
        ))}
      </tbody>
    </table>
  );
}
