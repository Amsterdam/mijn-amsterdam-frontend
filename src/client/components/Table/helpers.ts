import { DisplayProps } from './TableV2';
import { omit } from '../../../universal/helpers/utils';

export function withOmitDisplayPropsForSmallScreens<T extends object>(
  displayProps: DisplayProps<T>,
  omitKeysForSmallScreens: Array<keyof T>
) {
  return {
    ...displayProps,
    smallscreen: omit(displayProps, omitKeysForSmallScreens),
  };
}

export function getDisplayPropsForScreenSize<T extends object>(
  displayProps: DisplayProps<T>,
  isSmallScreen: boolean
) {
  return 'smallscreen' in displayProps && isSmallScreen
    ? displayProps.smallscreen
    : omit(displayProps, ['smallscreen']);
}
