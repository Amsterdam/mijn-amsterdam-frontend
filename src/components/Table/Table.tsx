import React, { ReactNode } from 'react';
import styles from './Table.module.scss';
import { entries } from 'helpers/App';
import Linkd from 'components/Button/Button';
import { Unshaped } from 'App.types';
import classnames from 'classnames';

export function addTitleLinkComponent(
  items: Unshaped[],
  titleKey: string = 'title'
) {
  return items.map((item: any) => {
    return {
      ...item,
      [titleKey]: <Linkd href={item.link.to}>{item[titleKey]}</Linkd>,
    };
  });
}

export interface TableProps {
  items: Unshaped[];
  className?: string;
  titleKey?: string;
  displayProps?: { [key: string]: string | number | ReactNode };
}

export default function Table({
  items,
  displayProps,
  titleKey = 'title',
  className,
}: TableProps) {
  const displayPropsFinal = !displayProps ? { [titleKey]: ' ' } : displayProps;
  const displayPropEntries = entries(displayPropsFinal).filter(
    ([key]) => key !== titleKey
  );
  const hasDisplayPropTableHeadingLabels = !!Object.keys(displayPropsFinal)
    .length;

  return (
    <table className={classnames(styles.Table, className)}>
      {hasDisplayPropTableHeadingLabels && (
        <thead>
          <tr className={styles.TableRow}>
            {!!items[0] && titleKey in items[0] && (
              <th className={styles.DisplayProp}>
                {displayPropsFinal[titleKey] || ' '}
              </th>
            )}
            {displayPropEntries.map(([key, label]) => (
              <th key={`th-${key}`} className={styles.DisplayProp}>
                {label}
              </th>
            ))}
          </tr>
        </thead>
      )}
      <tbody>
        {items.map((item: Unshaped, index) => (
          <tr
            key={item.id || `${titleKey}-${index}`}
            className={styles.TableRow}
          >
            {titleKey in item && (
              <td className={styles.DisplayPropTitle}>{item[titleKey]}</td>
            )}
            {displayPropEntries.map(([key, label]) => (
              <td key={`td-${key}`} className={styles.DisplayProp}>
                <span className={styles.DisplayPropLabel}>{label}:</span>
                {item[key] || <span>&mdash;</span>}
              </td>
            ))}
          </tr>
        ))}
      </tbody>
    </table>
  );
}
