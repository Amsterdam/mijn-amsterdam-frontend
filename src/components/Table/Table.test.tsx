import React from 'react';
import { shallow } from 'enzyme';
import Table from './Table';

it('Renders without crashing', () => {
  shallow(<Table />);
});
