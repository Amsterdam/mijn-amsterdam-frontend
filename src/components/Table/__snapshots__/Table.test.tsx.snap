// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Table /> Renderd differently without $titleKey and $displayProps in data object 1`] = `"<table class=\\"Table\\"><thead><tr class=\\"TableRow\\"></tr></thead><tbody><tr class=\\"TableRow\\"></tr><tr class=\\"TableRow\\"></tr></tbody></table>"`;

exports[`<Table /> Renderd differently without $titleKey and including $displayProps in data object 1`] = `"<table class=\\"Table\\"><thead><tr class=\\"TableRow\\"><th class=\\"DisplayProp\\">Datum</th></tr></thead><tbody><tr class=\\"TableRow\\"><td class=\\"DisplayProp\\"><span class=\\"DisplayPropLabel\\">Datum:</span>2019-08-06</td></tr><tr class=\\"TableRow\\"><td class=\\"DisplayProp\\"><span class=\\"DisplayPropLabel\\">Datum:</span>2013-08-06</td></tr></tbody></table>"`;

exports[`<Table /> Renders the default 1`] = `"<table class=\\"Table\\"><thead><tr class=\\"TableRow\\"><th class=\\"DisplayProp\\"> </th></tr></thead><tbody><tr class=\\"TableRow\\"><td class=\\"DisplayPropTitle\\">Title 1</td></tr><tr class=\\"TableRow\\"><td class=\\"DisplayPropTitle\\">Foo the world</td></tr></tbody></table>"`;

exports[`<Table /> Uses a different title key 1`] = `"<table class=\\"Table\\"><thead><tr class=\\"TableRow\\"><th class=\\"DisplayProp\\"> </th></tr></thead><tbody><tr class=\\"TableRow\\"><td class=\\"DisplayPropTitle\\">Title 1</td></tr><tr class=\\"TableRow\\"><td class=\\"DisplayPropTitle\\">Foo the world</td></tr></tbody></table>"`;
